shiny/express/_recall_context.py

```diff
@@ -0,0 +1,78 @@
+from __future__ import annotations
+
+import functools
+import sys
+from types import TracebackType
+from typing import Callable, Generic, Mapping, Optional, Type, TypeVar
+
+from htmltools import HTML, Tag, Tagifiable, TagList, tags
+
+from .._typing_extensions import ParamSpec
+
+P = ParamSpec("P")
+R = TypeVar("R")
+U = TypeVar("U")
+
+
+class RecallContextManager(Generic[R]):
+    def __init__(
+        self,
+        fn: Callable[..., R],
+        *,
+        default_page: RecallContextManager[Tag] | None = None,
+        args: tuple[object, ...] | None = None,
+        kwargs: Mapping[str, object] | None = None,
+    ):
+        self.fn = fn
+        self.default_page = default_page
+        if args is None:
+            args = tuple()
+        if kwargs is None:
+            kwargs = {}
+        self.args: list[object] = list(args)
+        self.kwargs: dict[str, object] = dict(kwargs)
+
+    def append_arg(self, value: object):
+        if isinstance(value, (Tag, TagList, Tagifiable)):
+            self.args.append(value)
+        elif hasattr(value, "_repr_html_"):
+            self.args.append(HTML(value._repr_html_()))  # pyright: ignore
+        else:
+            # We should NOT end up here for objects that were `def`ed, because they
+            # would already have been filtered out by _display_decorator_function_def().
+            # This is only for other kinds of expressions, the kind which would normally
+            # be printed at the console.
+            if value is not None:
+                self.args.append(tags.pre(repr(value)))
+
+    def __enter__(self) -> None:
+        if self.default_page is not None:
+            from . import _run
+
+            _run.replace_top_level_recall_context_manager(self.default_page)
+
+        self._prev_displayhook = sys.displayhook
+        # Collect each of the "printed" values in the args list.
+        sys.displayhook = self.append_arg
```

Comment:
We should consider just replacing `sys.displayhook` once, at process start time, with a smart version of our own. And then have that smart version look at a different variable, backed by a contextvar, to decide where things actually go. This would let us keep "displayhooks" straight even when using async logic.